File: admin/python/test/test_endpoints.py
"""
Unit tests for endpoints module (endpoints.py)
Tests all 5 endpoints: Users, Images, Operations, Jobs, Health
"""
import unittest
import os
import sys
import json
from unittest.mock import Mock, MagicMock, patch

sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

class TestUsersEndpoint(unittest.TestCase):
    """Test Users endpoint"""
    
    @patch('endpoints.cache')
    @patch('endpoints.db')
    def test_users_endpoint_cache_hit(self, mock_db, mock_cache):
        """Test Users endpoint with cache hit"""
        from endpoints import UsersEndpoint
        
        # Mock cache hit
        cached_data = json.dumps({
            'users': [{'id': 1, 'username': 'test', 'email': 'test@test.com'}],
            'cached': True,
            'cache_ttl': 300
        })
        mock_cache.get.return_value = cached_data
        
        endpoint = UsersEndpoint()
        result = endpoint.get()
        
        # Verify cache was checked
        mock_cache.get.assert_called_once()
        
        # Verify database was not queried
        mock_db.execute_query.assert_not_called()
        
        # Verify response
        response = json.loads(result)
        self.assertTrue(response['cached'])
        self.assertEqual(len(response['users']), 1)
    
    @patch('endpoints.cache')
    @patch('endpoints.db')
    def test_users_endpoint_cache_miss(self, mock_db, mock_cache):
        """Test Users endpoint with cache miss"""
        from endpoints import UsersEndpoint
        
        # Mock cache miss
        mock_cache.get.return_value = None
        
        # Mock database query result
        mock_db.execute_query.return_value = [
            {'id': 1, 'username': 'user1', 'email': 'user1@test.com',
             'created_at': '2024-01-01', 'image_count': 5, 'job_count': 3}
        ]
        
        endpoint = UsersEndpoint()
        result = endpoint.get()
        
        # Verify cache was checked
        mock_cache.get.assert_called_once()
        
        # Verify database was queried
        mock_db.execute_query.assert_called_once()
        
        # Verify cache was updated
        mock_cache.set.assert_called_once()
        
        # Verify response structure
        response = json.loads(result)
        self.assertFalse(response['cached'])
        self.assertEqual(response['cache_ttl'], 300)

class TestImagesEndpoint(unittest.TestCase):
    """Test Images endpoint"""
    
    @patch('endpoints.cache')
    @patch('endpoints.db')
    def test_images_endpoint_structure(self, mock_db, mock_cache):
        """Test Images endpoint returns correct structure"""
        from endpoints import ImagesEndpoint
        
        mock_cache.get.return_value = None
        mock_db.execute_query.return_value = [
            {
                'id': 1,
                'user_id': 1,
                'username': 'user1',
                'filename': 'test.jpg',
                'size': 1024,
                'content_type': 'image/jpeg',
                'uploaded_at': '2024-01-01'
            }
        ]
        
        endpoint = ImagesEndpoint()
        result = endpoint.get()
        
        response = json.loads(result)
        self.assertIn('images', response)
        self.assertIn('cached', response)
        self.assertEqual(len(response['images']), 1)
        
        image = response['images'][0]
        self.assertIn('id', image)
        self.assertIn('filename', image)
        self.assertIn('user_id', image)

class TestOperationsEndpoint(unittest.TestCase):
    """Test Operations endpoint"""
    
    @patch('endpoints.cache')
    @antml:parameter name="db')
    def test_operations_endpoint_structure(self, mock_db, mock_cache):
        """Test Operations endpoint returns statistics"""
        from endpoints import OperationsEndpoint
        
        mock_cache.get.return_value = None
        mock_db.execute_query.return_value = [
            {
                'operation': 'resize',
                'total': 100,
                'completed': 95,
                'failed': 3,
                'pending': 2,
                'avg_time_ms': 150.5
            }
        ]
        
        endpoint = OperationsEndpoint()
        result = endpoint.get()
        
        response = json.loads(result)
        self.assertIn('operations', response)
        self.assertEqual(len(response['operations']), 1)
        
        op = response['operations'][0]
        self.assertEqual(op['operation'], 'resize')
        self.assertEqual(op['total'], 100)
        self.assertIsInstance(op['avg_time_ms'], (int, float))

class TestJobsEndpoint(unittest.TestCase):
    """Test Jobs endpoint"""
    
    @patch('endpoints.cache')
    @patch('endpoints.db')
    def test_jobs_endpoint_structure(self, mock_db, mock_cache):
        """Test Jobs endpoint returns jobs and queue stats"""
        from endpoints import JobsEndpoint
        
        mock_cache.get.return_value = None
        
        # Mock two queries: jobs list and queue stats
        mock_db.execute_query.side_effect = [
            [  # Jobs list
                {
                    'id': 1,
                    'user_id': 1,
                    'username': 'user1',
                    'operation_type': 'resize',
                    'status': 'completed',
                    'created_at': '2024-01-01',
                    'completed_at': '2024-01-01'
                }
            ],
            [  # Queue stats
                {
                    'pending': 5,
                    'processing': 2,
                    'completed': 100,
                    'failed': 3
                }
            ]
        ]
        
        endpoint = JobsEndpoint()
        result = endpoint.get()
        
        response = json.loads(result)
        self.assertIn('jobs', response)
        self.assertIn('queue_stats', response)
        
        # Verify two queries were made
        self.assertEqual(mock_db.execute_query.call_count, 2)

class TestHealthEndpoint(unittest.TestCase):
    """Test Health endpoint"""
    
    @patch('endpoints.cache')
    @patch('endpoints.db')
    def test_health_endpoint_all_healthy(self, mock_db, mock_cache):
        """Test Health endpoint when all services healthy"""
        from endpoints import HealthEndpoint
        
        # Mock successful connections
        mock_db.execute_query.return_value = [{'result': 1}]
        mock_cache.redis_client.ping.return_value = True
        
        endpoint = HealthEndpoint()
        result = endpoint.get()
        
        response = json.loads(result)
        self.assertEqual(response['status'], 'healthy')
        self.assertTrue(response['database'])
        self.assertTrue(response['cache'])
    
    @patch('endpoints.cache')
    @patch('endpoints.db')
    def test_health_endpoint_db_failure(self, mock_db, mock_cache):
        """Test Health endpoint when database fails"""
        from endpoints import HealthEndpoint
        
        # Mock database failure
        mock_db.execute_query.side_effect = Exception("Connection failed")
        mock_cache.redis_client.ping.return_value = True
        
        endpoint = HealthEndpoint()
        result = endpoint.get()
        
        response = json.loads(result)
        self.assertEqual(response['status'], 'unhealthy')
        self.assertFalse(response['database'])

class TestEndpointIntegration(unittest.TestCase):
    """Integration tests for endpoints"""
    
    def test_all_endpoints_return_json(self):
        """Test that all endpoints return valid JSON"""
        from endpoints import UsersEndpoint, ImagesEndpoint, OperationsEndpoint, JobsEndpoint, HealthEndpoint
        
        endpoints = [
            UsersEndpoint(),
            ImagesEndpoint(),
            OperationsEndpoint(),
            JobsEndpoint(),
            HealthEndpoint()
        ]
        
        for endpoint in endpoints:
            with patch('endpoints.db'), patch('endpoints.cache'):
                # This test just verifies endpoints exist and can be instantiated
                self.assertIsNotNone(endpoint)

if __name__ == '__main__':
    unittest.main()

File: admin/cpp/test/test_endpoints.cpp
/**
 * Unit tests for endpoints module (endpoints.cpp/endpoints.h)
 * Tests all 5 endpoints: Users, Images, Operations, Jobs, Health
 */

#include <gtest/gtest.h>
#include <gmock/gmock.h>
#include "../endpoints.h"
#include "../db.h"
#include "../cache.h"

using ::testing::_;
using ::testing::Return;
using json = nlohmann::json;

class EndpointsTest : public ::testing::Test {
protected:
    void SetUp() override {
        setenv("DATABASE_URL", "postgresql://testuser:testpass@localhost:5432/testdb", 1);
        setenv("REDIS_URL", "redis://localhost:6379/0", 1);
    }
    
    void TearDown() override {
        unsetenv("DATABASE_URL");
        unsetenv("REDIS_URL");
    }
};

TEST_F(EndpointsTest, UsersEndpointExists) {
    // Test that UsersEndpoint class exists
    EXPECT_NO_THROW({
        UsersEndpoint endpoint;
    });
}

TEST_F(EndpointsTest, ImagesEndpointExists) {
    // Test that ImagesEndpoint class exists
    EXPECT_NO_THROW({
        ImagesEndpoint endpoint;
    });
}

TEST_F(EndpointsTest, OperationsEndpointExists) {
    // Test that OperationsEndpoint class exists
    EXPECT_NO_THROW({
        OperationsEndpoint endpoint;
    });
}

TEST_F(EndpointsTest, JobsEndpointExists) {
    // Test that JobsEndpoint class exists
    EXPECT_NO_THROW({
        JobsEndpoint endpoint;
    });
}

TEST_F(EndpointsTest, HealthEndpointExists) {
    // Test that HealthEndpoint class exists
    EXPECT_NO_THROW({
        HealthEndpoint endpoint;
    });
}

TEST_F(EndpointsTest, UsersEndpointReturnsJson) {
    // Test that Users endpoint returns valid JSON
    UsersEndpoint endpoint;
    
    DatabaseConnection db;
    CacheConnection cache;
    
    auto result = endpoint.handleRequest(db, cache);
    
    // Should return valid JSON string
    EXPECT_NO_THROW({
        json parsed = json::parse(result);
        EXPECT_TRUE(parsed.contains("users"));
        EXPECT_TRUE(parsed.contains("cached"));
    });
}

TEST_F(EndpointsTest, ImagesEndpointReturnsJson) {
    // Test that Images endpoint returns valid JSON
    ImagesEndpoint endpoint;
    
    DatabaseConnection db;
    CacheConnection cache;
    
    auto result = endpoint.handleRequest(db, cache);
    
    EXPECT_NO_THROW({
        json parsed = json::parse(result);
        EXPECT_TRUE(parsed.contains("images"));
        EXPECT_TRUE(parsed.contains("cached"));
    });
}

TEST_F(EndpointsTest, OperationsEndpointStructure) {
    // Test Operations endpoint response structure
    OperationsEndpoint endpoint;
    
    DatabaseConnection db;
    CacheConnection cache;
    
    auto result = endpoint.handleRequest(db, cache);
    
    EXPECT_NO_THROW({
        json parsed = json::parse(result);
        EXPECT_TRUE(parsed.contains("operations"));
        EXPECT_TRUE(parsed.contains("cached"));
        EXPECT_TRUE(parsed.contains("cache_ttl"));
        
        if (!parsed["operations"].empty()) {
            auto op = parsed["operations"][0];
            EXPECT_TRUE(op.contains("operation"));
            EXPECT_TRUE(op.contains("total"));
            EXPECT_TRUE(op.contains("avg_time_ms"));
        }
    });
}

TEST_F(EndpointsTest, JobsEndpointStructure) {
    // Test Jobs endpoint response structure
    JobsEndpoint endpoint;
    
    DatabaseConnection db;
    CacheConnection cache;
    
    auto result = endpoint.handleRequest(db, cache);
    
    EXPECT_NO_THROW({
        json parsed = json::parse(result);
        EXPECT_TRUE(parsed.contains("jobs"));
        EXPECT_TRUE(parsed.contains("queue_stats"));
        EXPECT_TRUE(parsed.contains("cached"));
        
        auto stats = parsed["queue_stats"];
        EXPECT_TRUE(stats.contains("pending"));
        EXPECT_TRUE(stats.contains("processing"));
        EXPECT_TRUE(stats.contains("completed"));
        EXPECT_TRUE(stats.contains("failed"));
    });
}

TEST_F(EndpointsTest, HealthEndpointStructure) {
    // Test Health endpoint response structure
    HealthEndpoint endpoint;
    
    DatabaseConnection db;
    CacheConnection cache;
    
    auto result = endpoint.handleRequest(db, cache);
    
    EXPECT_NO_THROW({
        json parsed = json::parse(result);
        EXPECT_TRUE(parsed.contains("status"));
        EXPECT_TRUE(parsed.contains("database"));
        EXPECT_TRUE(parsed.contains("cache"));
        
        // Status should be "healthy" or "unhealthy"
        std::string status = parsed["status"];
        EXPECT_TRUE(status == "healthy" || status == "unhealthy");
    });
}

TEST_F(EndpointsTest, EndpointsImplementCaching) {
    // Test that endpoints use caching
    UsersEndpoint endpoint;
    
    DatabaseConnection db;
    CacheConnection cache;
    
    // First call - cache miss
    auto result1 = endpoint.handleRequest(db, cache);
    json parsed1 = json::parse(result1);
    
    // Should have cached flag
    EXPECT_TRUE(parsed1.contains("cached"));
    EXPECT_TRUE(parsed1.contains("cache_ttl"));
}

int main(int argc, char **argv) {
    ::testing::InitGoogleTest(&argc, argv);
    return RUN_ALL_TESTS();
}
